openapi3filter/validate_request.go

```diff
@@ -97,58 +98,85 @@ func ValidateParameter(c context.Context, input *RequestValidationInput, paramet
-		if contentType == nil {
+	var (
+		req  = input.Request
+		data []byte
+	)
+
+	if req.Body != http.NoBody {
+		defer req.Body.Close()
+		var err error
+		if data, err = ioutil.ReadAll(req.Body); err != nil {
 			return &RequestError{
 				Input:       input,
 				RequestBody: requestBody,
-				Reason:      fmt.Sprintf("header 'Content-Type' has unexpected value: %q", inputMIME),
+				Reason:      "reading failed",
+				Err:         err,
 			}
 		}
-		schemaRef := contentType.Schema
-		if schemaRef != nil && isMediaTypeJSON(mediaType) {
-			schema := schemaRef.Value
-			body := req.Body
-			defer body.Close()
-			data, err := ioutil.ReadAll(body)
-			if err != nil {
-				return &RequestError{
-					Input:       input,
-					RequestBody: requestBody,
-					Reason:      "reading failed",
-					Err:         err,
-				}
-			}
+		// Put the data back into the input
+		req.Body = ioutil.NopCloser(bytes.NewReader(data))
+	}
 
-			// Put the data back into the input
-			req.Body = ioutil.NopCloser(bytes.NewReader(data))
+	if len(data) == 0 {
+		if requestBody.Required {
+			return &RequestError{Input: input, RequestBody: requestBody, Err: ErrInvalidRequired}
+		}
+		return nil
+	}
 
-			// Decode JSON
-			var value interface{}
-			if err := json.Unmarshal(data, &value); err != nil {
-				return &RequestError{
-					Input:       input,
-					RequestBody: requestBody,
-					Reason:      "decoding JSON failed",
-					Err:         err,
-				}
-			}
+	content := requestBody.Content
+	if len(content) == 0 {
+		// A request's body does not have declared content, so skip validation.
+		return nil
+	}
 
-			// Validate JSON with the schema
-			if err := schema.VisitJSON(value); err != nil {
-				return &RequestError{
-					Input:       input,
-					RequestBody: requestBody,
-					Reason:      "doesn't input the schema",
-					Err:         err,
-				}
-			}
+	inputMIME := req.Header.Get("Content-Type")
+	mediaType := parseMediaType(inputMIME)
+	if !isMediaTypeJSON(mediaType) {
+		// We support validation of JSON data only.
+		return nil
+	}
+
+	contentType := requestBody.Content[mediaType]
+	if contentType == nil {
+		return &RequestError{
+			Input:       input,
+			RequestBody: requestBody,
+			Reason:      fmt.Sprintf("header 'Content-Type' has unexpected value: %q", inputMIME),
+		}
+	}
+
+	schemaRef := contentType.Schema
+	if schemaRef == nil {
+		// A JSON schema that describes the received data is not declared, so skip validation.
+		return nil
+	}
+
+	// Decode JSON
+	var value interface{}
+	if err := json.Unmarshal(data, &value); err != nil {
+		return &RequestError{
+			Input:       input,
+			RequestBody: requestBody,
+			Reason:      "decoding JSON failed",
+			Err:         err,
+		}
+	}
+
+	// Validate JSON with the schema
+	if err := schemaRef.Value.VisitJSON(value); err != nil {
+		return &RequestError{
+			Input:       input,
+			RequestBody: requestBody,
+			Reason:      "doesn't input the schema",
```

Comment:
You mean `doesn't match the schema`?

Reply:
Yes, you are right. That's my bad. I will fix it.